### ISSUE TITLE
nnUNet model with 36 subjects (from ds004507 and spinegeneric)

### ISSUE THREAD

**Issue by tzebre** [ivadomed/model-spinal-rootlets#8]
## 1) Context
After training an nnUNet model on 18 subjects from [ds004507](https://openneuro.org/datasets/ds004507/versions/1.0.1) as described in [issue#7](https://github.com/ivadomed/model-spinal-rootlets/issues/7), predictions were made on 267 subjects from [spine-generic/data-multisubject](https://github.com/spine-generic/data-multi-subject). Some of the subjects were manually corrected. Following the merging of all subjects with manually reviewed labels, the dataset now includes 38 subjects, with 36 subjects in the training set and 2 subjects in the test set.

## 2) Results with binary labels
Results from nnUNet  :
<details>
<summary>inference_information.json</summary>

```
{
    "all_results": {
        "nnUNetTrainer__nnUNetPlans__3d_fullres": 0.6839309970081081
    },
    "best_model_or_ensemble": {
        "postprocessing_file": "/home/GRAMES.POLYMTL.CA/p118980/nvme/nnUNet_results/Dataset007_ds004507andspine/nnUNetTrainer__nnUNetPlans__3d_fullres/crossval_results_folds_0_1_2_3_4/postprocessing.pkl",
        "result_on_crossval_post_pp": 0.6839309970081081,
        "result_on_crossval_pre_pp": 0.6839309970081081,
        "selected_model_or_models": [
            {
                "configuration": "3d_fullres",
                "plans_identifier": "nnUNetPlans",
                "trainer": "nnUNetTrainer"
            }
        ],
        "some_plans_file": "/home/GRAMES.POLYMTL.CA/p118980/nvme/nnUNet_results/Dataset007_ds004507andspine/nnUNetTrainer__nnUNetPlans__3d_fullres/crossval_results_folds_0_1_2_3_4/plans.json"
    },
    "considered_models": [
        {
            "configuration": "3d_fullres",
            "plans": "nnUNetPlans",
            "trainer": "nnUNetTrainer"
        }
    ],
    "dataset_name_or_id": "Dataset007_ds004507andspine",
    "ensembling_allowed": true,
    "folds": [
        0,
        1,
        2,
        3,
        4
    ]
}
```
</details>
Dice score on sub-brnoUhb01 (data-multi-subject, not in training data): 0.771706

There can be 3 slices types of prediction results :
- True Positive and True Negative, 100% similarity ground truth vs predicted.
- Slice Positive (SP), 1 or + voxel similarity (not 100%) ground truth vs predicted. 
- False Positive and False Negative, only ground truth or predicted have voxel labeled. 

We can define new evaluation metrics: 
- Z-axis F1 score:  <img width="100" alt="Capture d’écran 2023-07-10 à 15 03 21" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/d3391491-ddc5-4879-874d-c866ca91215b">
- Mean common F1 score: <img width="222" alt="Capture d’écran 2023-07-10 à 15 08 17" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/57fcae0c-7f1b-4bb8-a44c-1adbcee2ee54">

With sub-brnoUhb01 this scores are:
- Z-axis F1 = 0.9017
- Mean common F1: 0.721

I made [a code](https://github.com/ivadomed/model-spinal-rootlets/blob/main/utilities/dice_y_value_soft.py) to compute these metrics. The code also creates 3 PDF files (Slice positive, False negative, False Positive) to compare ground truth vs prediction vs real image.
> Green : TP, Red: FP, Orange: FN, Black : TN

**Slice Positive**:
<img  width="349" alt="SP pdf" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/ae252521-1391-4dcf-856a-b7856c78c472">
**False Positive**:
<img width="312" alt="FP pdf" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/ca06d5bb-d9a1-4da7-977c-226648987cf7">
**False Negative**:
<img width="295" alt="FN pdf" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/c3169283-e44d-46fe-bc89-2cdddd70fbf4">

## 3) Results with levels specific labels
5 Images dropped from the dataset due to manual labeling difficulties.
In this part, the goal was to train a nnUNet model with voxel values specific to the spinal level. 
31 subjects from the previous dataset were labeled with voxel values depending on the spinal level (on fsleyes). 

To compare with the previous results, we binarized the labels and obtained the following scores with sub-brnoUhb01:
- Z-axis F1 = 0.842 (vs 0.9017)
- Mean common F1: 0.758 (vs 0.721) 

By levels:
- 2 
<img width="245" alt="Capture d’écran 2023-07-11 à 16 25 56" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/488da9a8-d69a-450c-a2ff-921d4a39823c"> 
<img width="400" alt="Capture d’écran 2023-07-12 à 09 34 55" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/f6f9ec16-d60e-472d-9f46-2054102c274c">

***

- 3
<img width="254" alt="Capture d’écran 2023-07-11 à 16 26 05" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/dd959b90-21cd-45e1-9f59-4aa1aa7e0b8e">
<img width="400" alt="Capture d’écran 2023-07-12 à 09 36 07" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/7abddc0c-0c46-49f6-b4de-595ffc11b5d2">

***

- 4
<img width="249" alt="Capture d’écran 2023-07-11 à 16 27 07" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/ee0c3e60-5c30-499d-81aa-6436e515ca9f">
<img width="400" alt="Capture d’écran 2023-07-12 à 09 36 23" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/9f0d6124-e088-445c-a31b-db5b20bf5418">

***

- 5
<img width="253" alt="Capture d’écran 2023-07-11 à 16 27 16" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/19327324-7423-4a9b-a653-17a998da3fea">
<img width="400" alt="Capture d’écran 2023-07-12 à 09 36 37" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/cfda033d-66db-4209-a9d3-df022281f860">

***

- 6
<img width="252" alt="Capture d’écran 2023-07-11 à 16 27 59" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/c39e3ba4-0821-4394-aaf8-be6fba6f89fc">
<img width="400" alt="Capture d’écran 2023-07-12 à 09 36 54" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/fd76a2bd-1cda-4cc0-8b95-76fca1ea561b">

***

- 7 
<img width="253" alt="Capture d’écran 2023-07-11 à 16 28 12" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/5c50942e-162e-4f2a-b922-d658660c0a53">
<img width="400" alt="Capture d’écran 2023-07-12 à 09 37 20" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/cdfc102c-c911-4150-93be-f1213d00c3a9">

***

- 8
<img width="254" alt="Capture d’écran 2023-07-11 à 16 30 13" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/3c310a43-f4ad-4aa8-abdd-6b9ca3151a89">
<img width="400" alt="Capture d’écran 2023-07-12 à 09 37 38" src="https://github.com/ivadomed/model-spinal-rootlets/assets/58122097/3340081d-bde5-4cb4-a550-0178857661d6">

**Comment by valosekj** [comment]
Follow-up models: https://github.com/ivadomed/model-spinal-rootlets/pull/23
